Model/Service/Index.php

```diff
@@ -0,0 +1,286 @@
+        $this->nostoSyncService = $nostoSyncService;
+    }
+
+    /**
+     * Handles only the first step of indexing
+     * Create one if row does not exits
+     * Else set row to dirty
+     *
+     * @param ProductCollection $collection
+     * @param Store $store
+     * @throws NostoException
+     */
+    public function invalidateOrCreate(ProductCollection $collection, Store $store)
+    {
+        $this->startBenchmark(
+            self::BENCHMARK_NAME_INVALIDATE,
+            self::BENCHMARK_BREAKPOINT_INVALIDATE
+        );
+        $collection->setPageSize(self::PRODUCT_DATA_BATCH_SIZE);
+        $iterator = new Iterator($collection);
+        $iterator->each(function (Product $item) use ($store) {
+            $this->updateOrCreateDirtyEntity($item, $store);
+            $this->tickBenchmark(self::BENCHMARK_NAME_INVALIDATE);
+        });
+        $this->logBenchmarkSummary(self::BENCHMARK_NAME_INVALIDATE, $store);
+    }
+
+    /**
+     * @param Product $product
+     * @param Store $store
+     * @return ProductIndexInterface|null
+     */
+    public function updateOrCreateDirtyEntity(Product $product, Store $store)
+    {
+        $indexedProduct = $this->indexRepository->getByProductIdAndStoreId($product->getId(), $store->getId());
+        try {
+            if ($indexedProduct instanceof ProductIndexInterface) {
+                $indexedProduct->setIsDirty(true);
+                $indexedProduct->setUpdatedAt($this->magentoTimeZone->date());
+            } else {
+                /* @var Product $fullProduct */
+                $fullProduct = $this->loadMagentoProduct($product->getId(), $store->getId());
+                $indexedProduct = $this->indexBuilder->build($fullProduct, $store);
+                $indexedProduct->setIsDirty(false);
+            }
+            $this->indexRepository->save($indexedProduct);
+            return $indexedProduct;
+        } catch (\Exception $e) {
+            $this->getLogger()->exception($e);
+            return null;
+        }
+    }
+
+    /**
+     * @param NostoIndexCollection $collection
+     * @param Store $store
+     * @throws NostoException
+     * @throws MemoryOutOfBoundsException
+     */
+    public function rebuildDirtyProducts(NostoIndexCollection $collection, Store $store)
+    {
+        $this->startBenchmark(
+            self::BENCHMARK_NAME_REBUILD,
+            self::BENCHMARK_BREAKPOINT_REBUILD
+        );
+        $collection->setPageSize(self::PRODUCT_DATA_BATCH_SIZE);
+        $iterator = new Iterator($collection);
+        $iterator->each(function (NostoProductIndex $item) {
+            if ($item->getIsDirty() === NostoProductIndex::DB_VALUE_BOOLEAN_TRUE) {
+                $this->rebuildDirtyProduct($item);
+                $this->tickBenchmark(self::BENCHMARK_NAME_REBUILD);
+            }
+            $this->checkMemoryConsumption('product rebuild');
+        });
+        $this->logBenchmarkSummary(self::BENCHMARK_NAME_REBUILD, $store);
```

Comment:
This should be in a finally block.

Reply:
The try-catch is inside a foreach loop. And logging is done at the very end of the loop.